lib/einhorn/command.rb

```diff
@@ -116,39 +116,65 @@ def self.signal_all(signal, children=nil, record=true)
           end
           spec[:signaled].add(signal)
+          spec[:last_signaled_at] = Time.now
         end
 
         begin
           Process.kill(signal, child)
         rescue Errno::ESRCH
+              Einhorn.log_debug("Attempted to #{signal} child #{child.inspect} but the process does not exist", :upgrade)
         else
           signaled[child] = spec
         end
       end
 
       if Einhorn::State.signal_timeout && record
         Einhorn::Event::Timer.open(Einhorn::State.signal_timeout) do
-          children.each do |child|
-            spec = Einhorn::State.children[child]
-            next unless spec # Process is already dead and removed by mourn
-            signaled_spec = signaled[child]
-            next unless signaled_spec # We got ESRCH when trying to signal
-            if spec[:spinup_time] != signaled_spec[:spinup_time]
-              Einhorn.log_info("Different spinup time recorded for #{child} after #{Einhorn::State.signal_timeout}s. This probably indicates a PID rollover.", :upgrade)
-              next
-            end
-
-            Einhorn.log_info("Child #{child.inspect} is still active after #{Einhorn::State.signal_timeout}s. Sending SIGKILL.")
-            begin
-              Process.kill('KILL', child)
-            rescue Errno::ESRCH
+          if Einhorn::State.config[:resignal_timeout]
+            resignal_children
+          else
+            children.each do |child|
+              spec = Einhorn::State.children[child]
+              next unless spec # Process is already dead and removed by mourn
+              signaled_spec = signaled[child]
+              next unless signaled_spec # We got ESRCH when trying to signal
+              if spec[:spinup_time] != signaled_spec[:spinup_time]
+                Einhorn.log_info("Different spinup time recorded for #{child} after #{Einhorn::State.signal_timeout}s. This probably indicates a PID rollover.", :upgrade)
+                next
+              end
+
+              Einhorn.log_info("Child #{child.inspect} is still active after #{Einhorn::State.signal_timeout}s. Sending SIGKILL.")
+              begin
+                Process.kill('KILL', child)
+              rescue Errno::ESRCH
+              end
+              spec[:signaled].add('KILL')
             end
-            spec[:signaled].add('KILL')
           end
+
+          "Successfully sent #{signal}s to #{signaled.length} processes: #{signaled.keys}"
         end
       end
+    end
 
-      "Successfully sent #{signal}s to #{signaled.length} processes: #{signaled.keys}"
+    def self.resignal_children
+      Einhorn::State.children.select{|_, c| c[:signaled] && c[:signaled].length > 0}.each do |pid, child|
+        next unless child[:last_signaled_at]
+
+        now = Time.now
+        expires_at = child[:last_signaled_at] + Einhorn::State.signal_timeout
+        next unless now >= expires_at
```

Comment:
I guess this should set up a new timer that fires at `expires_at`? Or is the assumption that this is set up already?

Reply:
Hm, interesting idea! Could definitely set up a new timer here. I was more going for "eventual consistency" where it would just clean up successfully on the next upgrade.